server/app/mutations/grid_services/common.rb

```diff
@@ -84,6 +86,30 @@ def build_grid_service_secrets(existing_secrets)
       service_secrets
     end
 
+    def build_service_volumes()
+      service_volumes = []
+      self.volumes.each do |vol|
+        vol_spec = parse_volume(vol)
+        if vol_spec[:volume]
+          # Named volume, try to find the proper mapping for external volumes
+          # NULL stack services don't have revs, for those just use the "global" name
+          volume_name = vol_spec[:volume]
+          if self.stack.latest_rev
+            stack_volume = self.stack.latest_rev.volumes.find {|v|
+              v['name'] == vol_spec[:volume]
+            }
+            # Use external volume definition if given
+            volume_name = stack_volume['external'] == true ? stack_volume['name'] : stack_volume.dig('external', 'name') || stack_volume['name']
```

Comment:
I can try to refactor it to be more clearer, but this essentially tries to use the given `external` volume name either with `external.name`or just the volume name when `external: true` is used.

The only option to take this functionality out from `Services` mutations is to pass some volume "mapping" hash into `Services` mutation. The issue I see there is that `Services` mutations are called from multiple sources, thus the mapping needs to be  generated in multiple places.